game-pipeline/MatchHandler.hs

```diff
@@ -0,0 +1,126 @@
+module Main where
+
+import Data.Aeson
+import qualified Data.ByteString.Char8 as B
+import qualified Data.ByteString.Lazy as LB
+import Data.Maybe
+import Options.Applicative
+import System.Exit
+import System.IO
+import System.Process
+import Types
+import Utils
+
+data Args = Args
+  { _gameEngine :: String,
+    _bots :: [String]
+  }
+  deriving (Show)
+
+argparser :: Parser Args
+argparser =
+  Args
+    <$> strOption
+      ( short 'g'
+          <> help "path to the game-engine"
+      )
+    <*> some
+      ( strOption
+          ( short 'b'
+              <> help "path to bots"
+          )
+      )
+
+opts :: ParserInfo Args
+opts =
+  info
+    (argparser <**> helper)
+    ( fullDesc
+        <> progDesc "Run a match"
+    )
+
+data Process = Process
+  { processHandle :: ProcessHandle,
+    stdIn :: Handle,
+    stdOut :: Handle,
+    stdErr :: Maybe Handle
+  }
+
+spawn :: String -> IO Process
+spawn path = do
+  (pstdin, pstdout, pstderr, ph) <-
+    createProcess
+      (shell path)
+        { std_in = CreatePipe,
+          std_out = CreatePipe
+        }
+  return $ Process ph (fromJust pstdin) (fromJust pstdout) pstderr
+
+getGameType :: GameEngineOutMessage Value -> Either String String
+getGameType (GameOStart i) = Right i
+getGameType _ = Left "Expected game engine message"
+
+liftEither :: Either String a -> IO a
+liftEither (Left errmsg) = error errmsg
+liftEither (Right a) = return a
+
+botInteract :: Process -> Value -> IO Value
+botInteract bot botin = do
+  flushedPutStrLnB (stdIn bot) (encode botin)
+  liftEither . eitherDecode . LB.fromStrict =<< B.hGetLine (stdOut bot)
+
+handleCommand :: Process -> [Process] -> GameEngineOutMessage Value -> IO ()
+handleCommand _ _ (GameEnd scores) = do
+  flushedPutStrLnB stdout $ encode (HGameEnd scores :: GameHistoryLine Value)
+  exitSuccess
+handleCommand gameEngine bots (DebugMessage msg) = do
+  flushedPutStrLnB stdout $ encode (HDebug msg :: GameHistoryLine Value)
+  gameLoop gameEngine bots
+handleCommand gameEngine bots (PlayerTurn playercmds) = do
+  botOuts <- mapM (\(p, c) -> botInteract (bots !! p) c) playercmds
+  flushedPutStrLnB stdout $
+    encode $
+      HPlayerResponses
+        ( zipWith
+            (curry (\(x, (p, _)) -> (p, x, True, "", "")))
+            botOuts
+            playercmds
+        )
```

Comment:
Hard to know what the arguments for `HPlayerResponses` mean. E.g. what are the `String`s for? Which one's which?

Reply:
I agree, I want to refactor the data types to use named fields as opposed to just a list of unnamed positional values. This will require a fairly large refactor of the data types so I kinda want to do it in a separate PR. I also want to do this refactor to make the datatypes more closely match the JSON contract to simplify the JSON parsers/encoders.